File: backend/services/payment_service.py
import os
from fastapi import HTTPException
from sqlalchemy.orm import Session
from sqlalchemy.sql import func
from models.delivery import Order
from models.payment import Payment

# Commission plateforme (10% par défaut)
PLATFORM_COMMISSION = float(os.getenv("PLATFORM_COMMISSION", "0.10"))

ORDER_STATUSES = ["en_attente", "payee", "expediee", "livree", "annulee"]


class PaymentService:

    def process_payment(self, db: Session, order_id: str, method: str = "carte",
                        card_number: str = None) -> dict:
        # SELECT FOR UPDATE pour éviter les doublons de paiement
        order = db.query(Order).filter(Order.id == order_id).with_for_update().first()
        if not order:
            raise HTTPException(status_code=404, detail="Commande introuvable")
        if order.status != "en_attente":
            raise HTTPException(
                status_code=400,
                detail="La commande ne peut pas être payée dans son état actuel",
            )

        # Simulation : carte commençant par 0000 → refusée
        if method == "carte" and card_number and card_number.startswith("0000"):
            raise HTTPException(status_code=402, detail="Paiement refusé (simulation)")

        amount = float(order.total_amount)
        commission = round(amount * PLATFORM_COMMISSION, 2)
        vendor_amount = round(amount - commission, 2)

        order.status = "payee"

        payment = Payment(
            order_id=order_id,
            amount=amount,
            commission=commission,
            vendor_amount=vendor_amount,
            method=method,
            status="validee",
        )
        db.add(payment)
        db.commit()

        return {
            "order_id": order_id,
            "status": "payee",
            "amount": amount,
            "commission": commission,
            "vendor_amount": vendor_amount,
            "method": method,
        }

    def get_order_status(self, db: Session, order_id: str) -> dict:
        row = (
            db.query(Order, Payment.method.label("payment_method"),
                     Payment.status.label("payment_status"))
            .outerjoin(Payment, Payment.order_id == Order.id)
            .filter(Order.id == order_id)
            .first()
        )
        if not row:
            raise HTTPException(status_code=404, detail="Commande introuvable")
        order, payment_method, payment_status = row
        return {
            "id": order.id,
            "status": order.status,
            "total_amount": float(order.total_amount),
            "created_at": str(order.created_at),
            "payment_method": payment_method,
            "payment_status": payment_status,
        }

    def update_order_status(self, db: Session, order_id: str, new_status: str) -> dict:
        if new_status not in ORDER_STATUSES:
            raise HTTPException(status_code=400, detail=f"Statut invalide : {new_status}")
        order = db.query(Order).filter(Order.id == order_id).first()
        if not order:
            raise HTTPException(status_code=404, detail="Commande introuvable")
        order.status = new_status
        db.commit()
        db.refresh(order)
        return {"id": order.id, "status": order.status}


payment_service = PaymentService()

File: backend/src/services/payment.service.js
const pool = require('../config/database');

// Statuts possibles d'une commande
const ORDER_STATUSES = ['en_attente', 'payee', 'expediee', 'livree', 'annulee'];

// Commission plateforme (10% par défaut)
const PLATFORM_COMMISSION = parseFloat(process.env.PLATFORM_COMMISSION) || 0.10;

class PaymentService {
  // Simulation de paiement
  async processPayment(orderId, { method = 'carte', cardNumber } = {}) {
    const client = await pool.connect();
    try {
      await client.query('BEGIN');

      const orderResult = await client.query(
        'SELECT * FROM orders WHERE id = $1 FOR UPDATE',
        [orderId]
      );
      const order = orderResult.rows[0];

      if (!order) {
        const err = new Error('Commande introuvable');
        err.status = 404;
        throw err;
      }
      if (order.status !== 'en_attente') {
        const err = new Error('La commande ne peut pas être payée dans son état actuel');
        err.status = 400;
        throw err;
      }

      // Simulation : on accepte toujours le paiement (sauf carte invalide)
      if (method === 'carte' && cardNumber && cardNumber.startsWith('0000')) {
        const err = new Error('Paiement refusé (simulation)');
        err.status = 402;
        throw err;
      }

      const commission = parseFloat(order.total_amount) * PLATFORM_COMMISSION;
      const vendorAmount = parseFloat(order.total_amount) - commission;

      await client.query(
        `UPDATE orders SET status = 'payee', updated_at = NOW() WHERE id = $1`,
        [orderId]
      );

      await client.query(
        `INSERT INTO payments (order_id, amount, commission, vendor_amount, method, status)
         VALUES ($1, $2, $3, $4, $5, 'validee')`,
        [orderId, order.total_amount, commission, vendorAmount, method]
      );

      await client.query('COMMIT');

      return {
        orderId,
        status: 'payee',
        amount: order.total_amount,
        commission,
        vendorAmount,
        method,
      };
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  }

  async getOrderStatus(orderId) {
    const result = await pool.query(
      `SELECT o.id, o.status, o.total_amount, o.created_at,
              p.method as payment_method, p.status as payment_status
       FROM orders o
       LEFT JOIN payments p ON p.order_id = o.id
       WHERE o.id = $1`,
      [orderId]
    );
    if (!result.rows[0]) {
      const err = new Error('Commande introuvable');
      err.status = 404;
      throw err;
    }
    return result.rows[0];
  }

  async updateOrderStatus(orderId, newStatus, userId, userRole) {
    if (!ORDER_STATUSES.includes(newStatus)) {
      const err = new Error(`Statut invalide : ${newStatus}`);
      err.status = 400;
      throw err;
    }

    const result = await pool.query(
      `UPDATE orders SET status = $1, updated_at = NOW()
       WHERE id = $2 RETURNING *`,
      [newStatus, orderId]
    );

    if (!result.rows[0]) {
      const err = new Error('Commande introuvable');
      err.status = 404;
      throw err;
    }
    return result.rows[0];
  }
}

module.exports = new PaymentService();
